[22.01] Drop uWSGI from scaling documentation (#13381)
galaxyproject/galaxy: doc/source/admin/scaling.md

    mapped to a handler tag such that all executions of that tool are handled by the tagged handler(s).
  * **default** - Any handlers without defined tags - aka "untagged handlers" - will handle executions of all tools
    not mapped to a specific handler ID or tag.
* **Webless Galaxy application** - The Galaxy application run as a standalone Python application with no web/ASGI server

## Application Servers

It is possible to run the Galaxy server in many different ways, including under different web application frameworks, or
as a standalone server with no web stack. Prior to the 18.01 release, Galaxy (by default)
used the [Python Paste][paste] web stack, and ran in a single process. Between the 18.01 release and the 22.01 release

> uWSGI was used as the default application server. Starting with the 22.01 release the default application server is

Reviewer: ```suggestion
[uWSGI](https://uwsgi-docs.readthedocs.io/) was used as the default application server. Starting with the 22.01 release the default application server is
```

> Gunicorn. For information about uWSGI in the Galaxy context please consult the version of this document that is

Reviewer: ```suggestion
[Gunicorn][Gunicorn]. For information about uWSGI in the Galaxy context please consult the version of this document that is
```

> appropriate to your Galaxy version.

Reviewer: ```suggestion
appropriate to your Galaxy release.
```

Gunicorn is able to serve ASGI applications. Galaxy can act as an ASGI web application since release 21.01,

> and we will drop support for being run as a WSGI application, and hence uWSGI compatibility in Galaxy release 22.05.

Reviewer: ```suggestion
and we will drop support for being run as a WSGI application via uWSGI in Galaxy release 22.05.
```

the configuration of the infrastructure on which you are deploying. In all cases, all Galaxy job features such as
[running on a cluster](cluster.md) are supported.

Although Gunicorn implements many features that were previously the responsibility of an upstream proxy server,
it is recommended to place a proxy server in front of Gunicorn and utilize it for all of its traditional
roles (serving static content, serving dataset downloads, etc.) as described in the [production
configuration](production.md) documentation.

### Gunicorn with jobs handled by web workers (default configuration)

Referred to in this documentation as the **all-in-one** strategy.

* Job handlers and web workers are the same processes and cannot be separated

> * A web worker that receives the job request from the UI/API will be the job handler for that job

Reviewer 2: ```suggestion
* A random web worker will be the job handler for that job
```

the configuration of the infrastructure on which you are deploying. In all cases, all Galaxy job features such as
[running on a cluster](cluster.md) are supported.

Although Gunicorn implements many features that were previously the responsibility of an upstream proxy server,
it is recommended to place a proxy server in front of Gunicorn and utilize it for all of its traditional
roles (serving static content, serving dataset downloads, etc.) as described in the [production
configuration](production.md) documentation.

### Gunicorn with jobs handled by web workers (default configuration)

Referred to in this documentation as the **all-in-one** strategy.

* Job handlers and web workers are the same processes and cannot be separated
* A web worker that receives the job request from the UI/API will be the job handler for that job

Under this strategy, jobs will be handled by Gunicorn workers. Having web processes handle jobs will negatively impact
UI/API performance.

> This is the default out-of-the-box configuration as of Galaxy Release 22.01.

Reviewer: ```suggestion
This is the default out-of-the-box configuration.
```

### uWSGI for web serving with Mules as job handlers

This strategy is deprecated and will be removed in Galaxy release 22.05.
You can consult the documentation for older versions of Galaxy for details.

If you're migrating Galaxy to 22.01 or newer we recommend you set up the

> **gunicorn + Webless** strategy below.

Reviewer: ```suggestion
**Gunicorn + Webless** strategy below.
```

* Additional job handlers can be added dynamically without reconfiguring/restarting Galaxy (19.01 or later)
* The recommended deployment strategy for production Galaxy instances

By default, handler assignment will occur using the **Database Transaction Isolation** or **Database SKIP LOCKED**
methods (see below). If the database used does not support this mechanism (in practice this should only apply
to sqlite before version 3.25, which is not at all recommended for production Galaxy server) a handler is randomly

> assigned by the web worker when the job is submitted via the UI/API, meaning that jobs may be assigned to unresponsive handlers.

Reviewer: This is covered in much more detail below, drop?

### uWSGI for web serving with Mules as job handlers

This strategy is deprecated and will be removed in Galaxy release 22.05.
You can consult the documentation for older versions of Galaxy for details.

> If you're migrating Galaxy to 22.01 or newer we recommend you set up the

Reviewer: ```suggestion
If you're upgrading Galaxy to release 22.01 or newer we recommend you set up the
```

**Gunicorn + Webless** strategy above.

## Job Handler Assignment Methods

Job handler assignment methods are configurable with the `assign_with`
attribute on the `<handlers>` tag in `job_conf.xml`.  The available methods are:

* **Database Transaction Isolation** (`db-transaction-isolation`, new in 19.01) - Jobs are assigned a handler by handlers selecting the unassigned
  job from the database using SQL transaction isolation, which uses database locks to guarantee that only one handler
  can select a given job. This occurs by the web worker that receives the tool execution request (via the UI or API)
  setting a new job's 'handler' column in the database to the configured tag/default (or `_default_` if no tag/default
  is configured). Handlers "listen" for jobs by selecting jobs from the database that match the handler tag(s) for which
  they are configured. `db-transaction-isolation` is the default assignment method if no handlers are defined,

>   or handlers are defined but no assign_with attribute is set on the `handlers` tag and *Database SKIP LOCKED* is not available.

Reviewer: ```suggestion
  or handlers are defined but no `assign_with` attribute is set on the `handlers` tag and *Database SKIP LOCKED* is not available.
```

* **Database SKIP LOCKED** (`db-skip-locked`, new in 19.01) - Jobs are assigned a handler by handlers selecting the unassigned job from
  the database using `SELECT ... FOR UPDATE SKIP LOCKED` on databases that support this query (see the next section for
  details). This occurs via the same process as *Database Transaction Isolation*, the only difference is the way in

>   which handlers query the database. This is the default if no handlers are defined, or handlers are defined but no assign_with attribute is set

Reviewer: ```suggestion
  which handlers query the database. This is the default if no handlers are defined, or handlers are defined but no `assign_with` attribute is set
```

### Choosing an Assignment Method

Prior to Galaxy 19.01, the most common deployment strategies (e.g. **uWSGI + Webless**) assigned handlers using what is
now (since 19.01) referred to as *Database Preassignment*.  Although still a fallback option when the database
does not support *Database SKIP LOCKED* or *Database Transaction Isolation*, preassignment has a few drawbacks:

* Web workers do not have a way to know whether a particular handler is alive when assigning that handler
* Jobs are not load balanced across handlers
* Changing the number of handlers requires changing `job_conf.xml` and restarting *all* Galaxy processes

The "database locking" methods (*Database SKIP LOCKED* and *Database Transaction Isolation*) were created to solve
these issues. The preferred method between the two options is *Database SKIP LOCKED*, but it requires PostgreSQL 9.5

> or newer, sqlite 3.25 or newer or MySQL 8.0 or newer (untested), or MariaDB 10.3 or newer (untested).

Reviewer: ```suggestion
or newer, SQLite 3.25 or newer or MySQL 8.0 or newer (untested), or MariaDB 10.3 or newer (untested).
```

If using an older database version, use *Database Transaction Isolation* instead. A detailed explanation of these database locking methods in PostgreSQL can be
found in the excellent [What is SKIP LOCKED for in PostgreSQL 9.5?][2ndquadrant-skip-locked] entry on the [2ndQuadrant
PostgreSQL Blog][2ndquadrant-blog].

The preferred assignment method is *Database SKIP LOCKED* or *Database Transaction Isolation*.

[2ndquadrant-skip-locked]: https://blog.2ndquadrant.com/what-is-select-skip-locked-for-in-postgresql-9-5/

> [2ndquadrant-blog]: https://blog.2ndquadrant.com/

Reviewer: ```suggestion
[2ndquadrant-skip-locked]: https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/
[2ndquadrant-blog]: https://www.2ndquadrant.com/en/blog/
```

## Configuration

### Gunicorn

We will only outline a few of Gunicorn's options, consult the [Gunicorn documentation](https://docs.gunicorn.org/en/latest/settings.html) for more.

> Note that by default Galaxy will use [gravity](https://github.com/galaxyproject/gravity/) to create a [supervisor](http://supervisord.org/) configuration that

Reviewer: Where is the supervisor configuration file created?

Author: You mean the command or the filesystem location ?

Reviewer: Not sure how this works, I supposed that gravity creates a `supervisord.conf` file somewhere?

Author: Ah, yes, that's in gravity's state_dir, by default in `database/gravity/supervisor`

> uses Gunicorn configuration values read from the `gravity` section of your `galaxy.yml` file.

Reviewer: ```suggestion
Note that by default Galaxy will use [gravity](https://github.com/galaxyproject/gravity/) to create a [supervisor](http://supervisord.org/) configuration
using Gunicorn configuration values read from the `gravity` section of your `galaxy.yml` file.
```

This is the preferred and out-of-the box way of configuring Gunicorn for serving Galaxy.

> If you are not using `./run.sh` for starting Galaxy or you would like to use another process manager

Reviewer: ```suggestion
If you are not using `./run.sh` for starting Galaxy or you would like to use another process manager,
```

> all of the Gunicorn configuration values can also directly be set on the command line.

Reviewer: ```suggestion
all the Gunicorn configuration values can also directly be set on the command line.
```

Configuration is performed in the `gravity` section of `galaxy.yml`. You will find that the default, if copied from
`galaxy.yml.sample`, is commented out. The default configuration options are provided to Gunicorn on the command line by
using `gravity` within the `run.sh` script.

> After making changes to the gravity section you always need to activate Galaxy's virtualenv and run `galaxyctl update`

Reviewer: ```suggestion
After making changes to the `gravity` section, you always need to activate Galaxy's virtualenv and run `galaxyctl update`
```

#### Common Gunicorn configuration

In `galaxy.yml`, define a `gravity` section. Shown below are the options common to all deployment strategies:

```yaml
gravity:
  app_server: gunicorn
  gunicorn:
    # listening options
    bind: '127.0.0.1:8080'
    # performance options
    workers: 1
    # Other options that will be passed to gunicorn
    gunicorn_extra_args:

```

Some of these options deserve explanation:

* `workers`: Controls the number of Galaxy application processes Gunicorn will spawn. Increased web performance can be
  attained by increasing this value. If Gunicorn is the only application on the server the number of CPUs * 2 + 1 is

>   a good starting value, 4 - 12 workers should be able to handle hundreds if not thousands of requests per second.

Reviewer: ```suggestion
  attained by increasing this value. If Gunicorn is the only application on the server, a good starting value is the number of CPUs * 2 + 1.
4-12 workers should be able to handle hundreds if not thousands of requests per second.
```

You can consult the Galaxy documentation for [Apache](apache.md) or [nginx](nginx.md)
for help with the proxy-side configuration.

By setting the `bind` option to a socket, `run.sh` will no longer automatically serve Galaxy via HTTP (since it is assumed that
you are setting a socket to serve Galaxy via a proxy server). If you wish to continue serving HTTP directly with Gunicorn
while using a socket, you can add an additional `--bind` argument via the `gunicorn_extra_args` option:

```yaml
gravity:
  app_server: gunicorn
  gunicorn:
    # listening options
    bind: '/srv/galaxy/var/gunicorn.sock'
    gunicorn_extra_args: '--bind 127.0.0.1:8080'
```

**Without a proxy server**:

It is strongly recommended to use a proxy server.

Gunicorn can be configured to serve HTTPS directly:

```yaml
  # listening options
  gunicorn:
    # listening options
    bind: '0.0.0.0:443'
    keyfile: server.key
    certfile: server.crt
```

See [Gunicorn's SSL documentation](https://docs.gunicorn.org/en/latest/settings.html#ssl) for more details.

To bind to ports < 1024 (e.g. if you want to bind to the standard HTTP/HTTPS ports 80/443), you must bind as the `root`

> user and drop privileges to the Galaxy user. However you are strongly encouraged to setup [Apache](apache.md) or [nginx](nginx.md) instead.

Reviewer: ```suggestion
user and drop privileges to the Galaxy user. However you are strongly encouraged to setup a proxy server
as described in the [production configuration](production.md) documentation.
```

## Configuration

### Gunicorn

We will only outline a few of Gunicorn's options, consult the [Gunicorn documentation](https://docs.gunicorn.org/en/latest/settings.html) for more.

Note that by default Galaxy will use [gravity](https://github.com/galaxyproject/gravity/) to create a [supervisor](http://supervisord.org/) configuration that
uses Gunicorn configuration values read from the `gravity` section of your `galaxy.yml` file.
This is the preferred and out-of-the box way of configuring Gunicorn for serving Galaxy.
If you are not using `./run.sh` for starting Galaxy or you would like to use another process manager
all of the Gunicorn configuration values can also directly be set on the command line.

> Configuration is performed in the `gravity` section of `galaxy.yml`. You will find that the default, if copied from

Reviewer 3: This is not setup yet. The [gravity readme](https://github.com/galaxyproject/gravity#configuration) refers to the *galaxy* section of `galaxy.yml` - I assume that's incorrect. ~~In any case, we need to add a `gravity` section to galaxy's config, right?~~ (I see, it can be manually added to `galaxy.yml`)

Author: Yeah, there's like 50 todos in here

## Configuration

### Gunicorn

> We will only outline a few of Gunicorn's options. Consult the Gunicorn documentation for more:

Reviewer: ```suggestion
We will only outline a few of Gunicorn's options, consult the [Gunicorn documentation](https://docs.gunicorn.org/en/latest/settings.html) for more.
```

> * [Configuring Gunicorn](https://docs.gunicorn.org/en/latest/settings.html)

Reviewer: ```suggestion
```

> Note that by default Galaxy will use [gravity](LINK) to create a [supervisor](LINK) configuration that

Reviewer: ```suggestion
Note that by default Galaxy will use [gravity](https://github.com/galaxyproject/gravity/) to create a [supervisor](http://supervisord.org/) configuration that
```

This is the preferred and out-of-the box way of configuring Gunicorn for serving Galaxy.
If you are not using `./run.sh` for starting Galaxy or you would like to use another process manager
all of the Gunicorn configuration values can also directly be set on the command line.

Configuration is performed in the `gravity` section of `galaxy.yml`. You will find that the default, if copied from
`galaxy.yml.sample`, is commented out. The default configuration options are provided to Gunicorn on the command line by
using `gravity` within the `run.sh` script.

After making changes to the gravity section you always need to activate Galaxy's virtualenv and run `galaxyctl update`

#### Common Gunicorn configuration

In `galaxy.yml`, define a `gravity` section. Shown below are the options common to all deployment strategies:

```yaml
gravity:
  app_server: gunicorn
  gunicorn:
    # TODO: CHANGE THIS IN GRAVITY, so we can use file descriptors
    # listening options
    bind: '127.0.0.1:8080'
    # performance options
    workers: 1
    # Other options that will be passed to gunicorn
    gunicorn_extra_args:

```

> Some of these options warrant explanation:

Reviewer: ```suggestion
Some of these options deserve explanation:
```

* `workers`: Controls the number of Galaxy application processes Gunicorn will spawn. Increased web performance can be
  attained by increasing this value. If Gunicorn is the only application on the server the number of CPUs * 2 + 1 is
  a good starting value, 4 - 12 workers should be able to handle hundreds if not thousands of requests per second.
* `gunicorn_extra_args`: You can specify additional arguments to pass to gunicorn here.

> Additional options are explained in the [uWSGI Minutiae](#uwsgi-minutiae) below.
> 

Reviewer: ```suggestion
```

   ``handler3`` will *only* handle executions of tool ``test1``.
```

> `run.sh` will start the gunicorn and job handler process(es), but if you are not using run.sh or the generated supervisor setup you will need to start the webless handler processes yourself. This is done on the command line like so:

Reviewer: ```suggestion
`run.sh` will start the gunicorn and job handler process(es), but if you are not using `run.sh` or the generated supervisor setup you will need to start the webless handler processes yourself. This is done on the command line like so:
```

</job_conf>
```

Note that we have defined a `<handlers>` section without any `<handler>` entries,
and we have explicitly configured the assignment method with `assign_with="db-skip-locked"`.

To let gravity know how many webless handler processes should be started set the number of processes to start in the `gravity:` section of galaxy.yml:

```yaml
gravity:
    handlers:
      handler:
        processes: 3
        pools:
          - job-handler
          - workflow-scheduler
      special:
        pools:

>           - job-handler.special

Reviewer: ```suggestion
  handlers:
    handler:
      processes: 3
      pools:
        - job-handler
        - workflow-scheduler
    special:
      pools:
        - job-handler.special
```

./scripts/galaxy-main -c config/galaxy.yml --server-name handler_3 --attach-to-pool job-handlers --attach-to-pool workflow-scheduler --daemonize
./scripts/galaxy-main -c config/galaxy.yml --server-name special_0 --attach-to-pool job-handlers.special --daemonize
```

In this example:

* `handler_0`, `handler_1` and `handler2` will handle tool executions that are not explicitly mapped to handlers
* `special_0` will handle tool executions that are mapped to the `special` handler tag

## gravity & galaxyctl

[Gravity](https://github.com/galaxyproject/gravity) is a management tool for Galaxy servers,
and is installed when you set up Galaxy.

It provides two executables, `galaxyctl`, which is used to manage the starting, stopping, and logging of Galaxy's various processes, and `galaxy`, which can be used to run a Galaxy server in the foreground.

These commands are available from within Galaxy's virtualenv, or you can install them globally.
If you have used the standard installation method for Galaxy by running `./run.sh` or executing
`./scripts/common_startup.sh`, a default directory has been configured in which gravity stores

> it's state. If you have installed Galaxy or gravity by another means you can use the `--state-dir` argument

Reviewer: ```suggestion
its state. If you have installed Galaxy or gravity by another means, you can use the `--state-dir` argument
```

> or the `GRAVITY_STATE_DIR` environment variable to control the state dir.

Reviewer: ```suggestion
or the `GRAVITY_STATE_DIR` environment variable to control the state directory.
```

In the following sections we assume you have correctly set up gravity and can use the `galaxyctl` command.

## Logging and daemonization

When running `./run.sh` the log output is shown on screen, and ^ctrl-c will stop all processes.
Galaxy can be started in the background by running `./run.sh --daemon`.

> Alternatively, you can control galaxy using the `galaxyctl` command provided by gravity.

Reviewer: ```suggestion
Alternatively, you can control Galaxy using the `galaxyctl` command provided by gravity.
```

[2022-02-17 16:31:27,035: DEBUG/MainProcess] | Consumer: Starting event loop
[2022-02-17 16:31:27,035: INFO/MainProcess] celery@MacBook-Pro-2.local ready.
[2022-02-17 16:31:27,035: DEBUG/MainProcess] basic.qos: prefetch_count->8

==> /Users/mvandenb/src/doc_test/database/gravity/log/celery-beat.log <==
[2022-02-17 16:29:04,023: DEBUG/MainProcess] beat: Ticking with max interval->5.00 minutes
[2022-02-17 16:29:04,024: DEBUG/MainProcess] beat: Waking up in 5.00 minutes.
No Galaxy config file found, running from current working directory: /Users/mvandenb/src/doc_test
[2022-02-17 16:31:26,818: DEBUG/MainProcess] Setting default socket timeout to 30
[2022-02-17 16:31:26,819: INFO/MainProcess] beat: Starting...
[2022-02-17 16:31:26,824: DEBUG/MainProcess] Current schedule:
<ScheduleEntry: prune-history-audit-table galaxy.celery.tasks.prune_history_audit_table() <freq: 1.00 hour>
<ScheduleEntry: celery.backend_cleanup celery.backend_cleanup() <crontab: 0 4 * * * (m/h/d/dM/MY)>
[2022-02-17 16:31:26,824: DEBUG/MainProcess] beat: Ticking with max interval->5.00 minutes
[2022-02-17 16:31:26,825: DEBUG/MainProcess] beat: Waking up in 5.00 minutes.
```

> More advanced logging options are described in the Galaxy [Logging Configuration documentation](config_logging)

Reviewer: ```suggestion
More advanced logging options are described in the Galaxy [Logging Configuration documentation](config_logging).
```

## Starting and Stopping

If you want to run your Galaxy server as a persistent service, you can include the `galaxy` script from Galaxy's
virtualenv in the configuration of your process manager (e.g. systemd). You can then continue using the `galaxyctl` command as usual
to start/stop/restart Galaxy or follow the logs.

### Transparent restarts

For zero-downtime restarts use the command `galaxyctl graceful`.

### Systemd

This is a sample config for systemd. More information on systemd.service environment

> settings can be found in the [documentation](http://0pointer.de/public/systemd-man/systemd.exec.html#Environment=)

Reviewer: ```suggestion
settings can be found in the [documentation](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Environment=)
```

./scripts/galaxy-main -c config/galaxy.yml --server-name special_0 --attach-to-pool job-handlers.special --daemonize
```

In this example:

* `handler_0`, `handler_1` and `handler2` will handle tool executions that are not explicitly mapped to handlers
* `special_0` will handle tool executions that are mapped to the `special` handler tag

## gravity & galaxyctl

[Gravity](https://github.com/galaxyproject/gravity) is a management tool for Galaxy servers,
and is installed when you set up Galaxy.

It provides two executables, `galaxyctl`, which is used to manage the starting, stopping, and logging of Galaxy's various processes, and `galaxy`, which can be used to run a Galaxy server in the foreground.

These commands are available from within Galaxy's virtualenv, or you can install them globally.
If you have used the standard installation method for Galaxy by running `./run.sh` or executing
`./scripts/common_startup.sh`, a default directory has been configured in which gravity stores
its state. If you have installed Galaxy or gravity by another means, you can use the `--state-dir` argument

> or the `GRAVITY_STATE_DIR` environment variable to control the state directory.

Reviewer 2: ```suggestion
or the `GRAVITY_STATE_DIR` environment variable to control the state directory. If a state dir is not specified, it defaults to `~/.config/galaxy-gravity`.
```

In the following sections we assume you have correctly set up gravity and can use the `galaxyctl` command.

## Logging and daemonization

When running `./run.sh` the log output is shown on screen, and ^ctrl-c will stop all processes.
Galaxy can be started in the background by running `./run.sh --daemon`.

Alternatively, you can control Galaxy using the `galaxyctl` command provided by gravity.

> After activating Galaxy's virtualenv environment you can start Galaxy in the background

Reviewer 2: ```suggestion
After activating Galaxy's virtual environment you can start Galaxy in the background
```

</job_conf>
```

To let gravity know how many webless handler processes should be started set the `job_handler_count` value in galaxy.yml:

```yaml
gravity:
  ...
  job_handler_count: 4
  ...
```

As with statically defined handlers, `run.sh` will start the process(es), but if you are not using `run.sh` or the generated supervisor config you will need to start the webless handler processes yourself. This is done on the command line like so (note the addition of the `--attach-to-pool`
option):

TODO: `job_handler_count` doesn't really make sense if you have multiple pools. Could be

```yaml
gravity:
  ...
  job_handler_count:
    _default_: 4
    special: 2
```

> ?

Author: If you have other idea let me know!

Reviewer 2: Maybe if it's an integer it's assumed to be `_default_` and if it's a dict it works as you've shown it here? Also doesn't handle workflow schedulers currently.

Author: Ah, that is a good point. We should also anticipate scaling the celery workers.
Maybe
```yaml
gravity:
    ...
    process_count:
        job_handler:
           _default_: 10
          special: 2
       workflow_handler: 3
       celery: 2
```
?

</job_conf>
```

To let gravity know how many webless handler processes should be started set the number of processes to start in the `gravity:` section of galaxy.yml:

```yaml
gravity:
  ...

>   process_count:

Author: Maybe it'd make more sense to control the process_count within their own key,
so 
```
gravity:
  ...
  job_handler:
    _default_: 
      processes: 4
    special: 2
  workflow_handler:
    processes: 2
  celery:
    concurrency: 2
    extra_args: '--debug'
    
```
I think that would look a little better.

Reviewer 2: `processes` as its own key under `job_handler` or `workflow_handler` doesn't distinguish it as "special" from a pool name (tag). How about:

```gravity:
  ...
  job_handler:
    _default_: 
      processes: 4
    special: 2
  workflow_handler: 2
  celery:
    concurrency: 2
    extra_args: '--debug'
```

Reviewer 2: Although how to specify the name template?

Author: For the name template we could do:

```yaml
  job_handler:
    _default_: 
      processes: 4
     name_template: the_template_{i}
    special: 2
  workflow_handler:
    _default_:
      processes: 2
```
But since the pools are the top level keys here we can't really have a construct to say that a handler should join multiple pools (and/or tags). Maybe we should make the first portion of the handler name the top level key ?

```yaml
gravity:
    handlers:
      handler:
        processes: 2
        name_template: >
          {name}{process}
        pools:
          - job-handler
          - workflow-scheduler
      handler1:
        processes: 1
        pools:
          - job-handler.special
      handler2:
        processes: 1
        pools:
          - job-handler
          - job-handler.special
```
(handler1 and handler2 are overrides here)

would be the equivalent of the old documentation for
> - `handler0` and `handler2` will handle tool executions that are not explicitly mapped to handlers
> - `handler1` and `handler2` will handle tool executions that are mapped to the `special` handler tag

(and handler 1 will only handle special, which we didn't spell out before)

Author: I've implemented the last example in https://github.com/galaxyproject/gravity/pull/28/commits/a00df1c671fdc01e3fbc42f64a851a45a37a1870

==> /Users/mvandenb/src/doc_test/database/gravity/log/celery-beat.log <==
[2022-02-17 16:29:04,023: DEBUG/MainProcess] beat: Ticking with max interval->5.00 minutes
[2022-02-17 16:29:04,024: DEBUG/MainProcess] beat: Waking up in 5.00 minutes.
No Galaxy config file found, running from current working directory: /Users/mvandenb/src/doc_test
[2022-02-17 16:31:26,818: DEBUG/MainProcess] Setting default socket timeout to 30
[2022-02-17 16:31:26,819: INFO/MainProcess] beat: Starting...
[2022-02-17 16:31:26,824: DEBUG/MainProcess] Current schedule:
<ScheduleEntry: prune-history-audit-table galaxy.celery.tasks.prune_history_audit_table() <freq: 1.00 hour>
<ScheduleEntry: celery.backend_cleanup celery.backend_cleanup() <crontab: 0 4 * * * (m/h/d/dM/MY)>
[2022-02-17 16:31:26,824: DEBUG/MainProcess] beat: Ticking with max interval->5.00 minutes
[2022-02-17 16:31:26,825: DEBUG/MainProcess] beat: Waking up in 5.00 minutes.
```

More advanced logging options are described in the Galaxy [Logging Configuration documentation](config_logging)

## Starting and Stopping

If you want to run your Galaxy server as a persistent service, you can include the `galaxy` script from Galaxy's

> virtualenv in your process manager configuration. You can then continue using the `galaxyctl` command as usual

Reviewer: ```suggestion
virtualenv in the configuration of your process manager (e.g. systemd). You can then continue using the `galaxyctl` command as usual
```

to start/stop/restart Galaxy or follow the logs.

> ## Transparent restarts

Reviewer: Maybe move this section after `### Systemd` ? Also, do you want it to be under `## Starting and Stopping`? If yes, you need an additional `#`.

Author: yep, that's a good idea.

Author: ```suggestion
### Transparent restarts
```
